File: tests/functional2/testlib/terminal_code_eater.py
# copy pasta from libutil-support's terminal-code-eater.cc

import enum
import dataclasses


class State(enum.Enum):
    ExpectESC = 1
    ExpectESCSeq = 2
    InCSIParams = 3
    InCSIIntermediates = 4
    InOSCParams = 5
    InOSCST = 6


@dataclasses.dataclass
class TerminalCodeEater:
    state: State = State.ExpectESC

    def feed(self, data: bytes) -> bytes:
        def is_param_char(char: int) -> bool:
            return 48 <= char <= 63

        def is_intermediate_char(char: int) -> bool:
            return 32 <= char <= 47

        def is_final_char(char: int) -> bool:
            return 64 <= char <= 126

        ret = bytearray()
        for c in data:
            match self.state:
                case State.ExpectESC:
                    match c:
                        case 0x1B:  # \e
                            self._transition(State.ExpectESCSeq)
                            continue
                        case 0xD:  # \r
                            continue
                    ret.append(c)
                case State.ExpectESCSeq:
                    match c:
                        case 0x5B:
                            # corresponds to CSI ('[')
                            self._transition(State.InCSIParams)
                            continue
                        case 0x5D:
                            # corresponds to OSC (']')
                            self._transition(State.InOSCParams)
                            continue
                        # FIXME(jade): whatever this was, we do not know how to
                        #  delimit it, so we just eat the next character and
                        #  keep going. Should we actually eat it?
                        case _:
                            self._transition(State.ExpectESC)
                            continue
                # https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_(Control_Sequence_Introducer)_sequences
                # A CSI sequence is: CSI [\x30-\x3f]* [\x20-\x2f]* [\x40-\x7e]
                #                        ^ params     ^ intermediates ^ final byte
                case State.InCSIParams:
                    if is_final_char(c):
                        self._transition(State.ExpectESC)
                        continue
                    if is_intermediate_char(c):
                        self._transition(State.InCSIIntermediates)
                        continue
                    if is_param_char(c):
                        continue
                    msg = f"Corrupt escape sequence, at {c:x}"
                    raise ValueError(msg)
                case State.InCSIIntermediates:
                    if is_final_char(c):
                        self._transition(State.ExpectESC)
                        continue
                    if is_intermediate_char(c):
                        continue
                    msg = f"Corrupt escape sequence in intermediates, at {c:x}"
                    raise ValueError(msg)
                # An OSC is OSC [\x20-\x7e]* ST per ECMA-48
                # where OSC is \x1b ] and ST is \x1b \.
                case State.InOSCParams:
                    # first part of ST
                    if c == 0x1B:
                        self._transition(State.InOSCST)
                        continue
                    # OSC sequences can be ended by BEL on old xterms
                    if c == 0x07:
                        self._transition(State.ExpectESC)
                        continue
                    if c < 0x20 or c == 0x7F:
                        msg = f"Corrupt OSC sequence, at {c:x}"
                        raise ValueError(msg)
                    # either way, eat it
                    continue
                case State.InOSCST:
                    # ST ends by \
                    if c == 0x5C:  # \
                        self._transition(State.ExpectESC)
                    elif c < 0x20 or c > 0x7E:
                        msg = f"Corrupt OSC sequence in ST, at {c:x}"
                        raise ValueError(msg)
                    else:
                        self._transition(State.InOSCParams)
                    continue

        return bytes(ret)

    def _transition(self, new_state: State):
        self.state = new_state


def eat_terminal_codes(s: bytes) -> bytes:
    return TerminalCodeEater().feed(s)

File: tests/unit/libutil-support/tests/terminal-code-eater.cc
// this file has a hissing snake twin in functional2/testlib/terminal_code_eater.py
#include "terminal-code-eater.hh"
#include "lix/libutil/escape-char.hh"
#include <assert.h>
#include <cstdint>
#include <iostream>

namespace nix {

static constexpr const bool DEBUG_EATER = false;

void TerminalCodeEater::feed(char c, std::function<void(char)> on_char)
{
    auto isParamChar = [](char v) -> bool { return v >= 0x30 && v <= 0x3f; };
    auto isIntermediateChar = [](char v) -> bool { return v >= 0x20 && v <= 0x2f; };
    auto isFinalChar = [](char v) -> bool { return v >= 0x40 && v <= 0x7e; };
    if constexpr (DEBUG_EATER) {
        std::cerr << "eater" << MaybeHexEscapedChar{c} << "\n";
    }

    switch (state) {
    case State::ExpectESC:
        switch (c) {
        case '\e':
            transition(State::ExpectESCSeq);
            return;
        // Just eat \r, since it is part of clearing a line
        case '\r':
            return;
        default:
            break;
        }
        if constexpr (DEBUG_EATER) {
            std::cerr << "eater uneat" << MaybeHexEscapedChar{c} << "\n";
        }
        on_char(c);
        break;
    case State::ExpectESCSeq:
        switch (c) {
        // CSI
        case '[':
            transition(State::InCSIParams);
            return;
        case ']':
            transition(State::InOSCParams);
            return;
        // FIXME(jade): whatever this was, we do not know how to delimit it, so
        // we just eat the next character and keep going
        default:
            transition(State::ExpectESC);
            return;
        }
        break;
    // https://en.wikipedia.org/wiki/ANSI_escape_code#CSI_(Control_Sequence_Introducer)_sequences
    // A CSI sequence is: CSI [\x30-\x3f]* [\x20-\x2f]* [\x40-\x7e]
    //                        ^ params     ^ intermediates ^ final byte
    case State::InCSIParams:
        if (isFinalChar(c)) {
            transition(State::ExpectESC);
            return;
        } else if (isIntermediateChar(c)) {
            transition(State::InCSIIntermediates);
            return;
        } else if (isParamChar(c)) {
            return;
        } else {
            // Corrupt escape sequence? Throw an assert, for now.
            // transition(State::ExpectESC);
            assert(false && "Corrupt terminal escape sequence");
            return;
        }
        break;
    case State::InCSIIntermediates:
        if (isFinalChar(c)) {
            transition(State::ExpectESC);
            return;
        } else if (isIntermediateChar(c)) {
            return;
        } else {
            // Corrupt escape sequence? Throw an assert, for now.
            // transition(State::ExpectESC);
            assert(false && "Corrupt terminal escape sequence in intermediates");
            return;
        }
        break;
    // An OSC is OSC [\x20-\x7e]* ST
    // where OSC is \x1b ] and ST is \x1b \.
    case State::InOSCParams:
        if (c == '\e') {
            // first part of ST
            transition(State::InOSCST);
        } else if (c == '\a') {
            // OSC sequences can be ended by BEL on old xterms
            transition(State::ExpectESC);
        } else if (c < 0x20 or c > 0x7e) {
            assert(false && "Corrupt OSC sequence");
        }
        // either way, eat it
        return;
    case State::InOSCST:
        // ST ends by \.
        if (c == '\\') {
            transition(State::ExpectESC);
        } else if (c < 0x20 || c == 0x7f) {
            assert(false && "Corrupt OSC sequence, in ST");
        } else {
            transition(State::InOSCParams);
        }
        return;
    }
}

void TerminalCodeEater::transition(State new_state)
{
    state = new_state;
}
};
